### ISSUE TITLE
zf_log_util - output callbacks (sinks) collection

### ISSUE THREAD

**Issue by wonder-mice** [wonder-mice/zf_log#9]
It would be useful to have most common output callbacks (sinks) implemented in a companion (utility) library. Things that come to mind:
* Sync file output callback
* Async file output callback

**Comment by wonder-mice** [comment]
Moving discussion here from https://github.com/wonder-mice/zf_log/issues/16:

> who implements output_open() and _close() , each individual "driver"?

I think so. Don't see any reason to generalize it even further, since set of initial parameters will be different for each output sink. And some sinks will want to have additional runtime configuration. Say:

```c
zf_log_output file_output_open(const char *const file, const size_t max_file_size);
void file_output_rotate(zf_log_output *output, const char *const backup_file);
void file_output_close(zf_log_output *output);
```

**Comment by fenugrec** [comment]
Agreed, and the caller knows which one it wants anyway so an extra abstraction wouldn't be super useful. So, xyz_output_open() initializes the output "driver" and returns a new "zf_log_output", does it also assign the output callback ? Or is there an extra zf_() call to do that at initialization ?

**Comment by wonder-mice** [comment]
I think xyz_output_open() must fully initialize `zf_log_output` structure. In the end, it's all about output callback and what it needs. In that case zf_log_output works as an opaque handler, only corresponding set of `xyz_output_` functions really knows what's there and how to work with that. Say, `arg` field can be set to malloc-ed buffer, that `xyz_output_close()` will free.

**Comment by fenugrec** [comment]
Hi, I'm looking at this again, did you have some work done already on this ?

**Comment by wonder-mice** [comment]
No.

**Comment by fenugrec** [comment]
Well, I decided to do something about this before the year is out. I'll start with a simple sync file IO output with a mutex for single-process, multi-thread logging. I will write native win (WriteFile + CriticalSection)  and *nix (write +  pthread_mutex) implementations, for now I won't change core zf_log components unless necessary. I'll submit a PR when I'm done.
